Draft of async API

lutovich opened the pull request:
PR adds initial version of async API for direct (only `bolt://` scheme, not `bolt+routing://`) driver.

Main API contact points:
 * `Response` interface that is a `Future` one can subscribe to with `ResponseListener`
 * `StatementResultCursor` which is an async counterpart of `StatementResult`; it is basically an async iterator with `#hasNext()` being `Response<Boolean> fetchAsync()` and `#next()` being `Record current()`
 * various overloads or `Session#runAsync()` to run queries asynchronously in auto-commit mode
 * `Session#beginTransactionAsync()` to start a transaction asynchronously
 * various overloads or `Transaction#runAsync()` to run queries asynchronously in the transaction
 * `Transaction#commitAsync()` and `Transaction#rollbackAsync()` to finish transaction with commit or rollback
 * `Session#closeAsync()` to close session asynchronously and release all resources

Implementation notes:
 * Netty is used internally for async IO; it is now specified as a dependency with scope `compile` but will eventually be shaded into the driver artifact
 * All internal futures & promises are built on top of corresponding Netty classes; all callbacks and future transformations execute in Netty event loop threads
 * Channels are pooled in a `FixedChannelPool` which adds two new config settings: `maxConnectionPoolSize` and `connectionAcquisitionTimeoutMillis`
 * Javadocs are missing right now
 * Packages will need some reshuffling
 * Simple back-pressure is available using channel auto-read functionality but it's turned off by default
 * Async API lives in parallel to sync API and uses separate network connections, pools and everything; this is temporary and eventually sync will be implemented on top of async

Not part of this PR:
 * Transaction functions with retries
 * Routing driver
 * Netty logging does not use driver logging
 * Protocol level back-pressure
 * Good test coverage

Related to https://github.com/neo4j/neo4j-java-driver/issues/102

lutovich committed:
Initial attempt to use Netty in connection layer

Added runtime dependency on netty-all. Introduced a dedicated package
for the experimental classes. New `NettyConnection` is able to
asynchronously send and receive all types of Bolt messages. All
classes are completely disconnected from the rest of the codebase.

lutovich committed:
Simple async connection with RUN, PULL_ALL and flush

Added separate netty based async connection. Made netty channel
connector return initialized (after INIT) connection futures.

lutovich committed:
Measure average query time, print less to sout

lutovich committed:
Simplified collectors

Collectors are basically response handlers that are queued at the same
time as outbound message in a separate queue. Later, when response
arrives next collector is popped from the queue to handle it.
Collector interface was quite wide and allowed processing parts of the
SUCCESS message metadata.

This commit turns `Collector` into `ResponseHandler` and makes it only
expose three methods to handle SUCCESS, FAILURE and RECORD messages.

lutovich committed:
SSL support in netty-based connection

Made it possible to pass `SecurityPlan` down to the Netty channel
bootstrap and add an SSL handler, if needed. This allows running
encrypted connection from driver to server.

Also fixed compilation problems related to `Collector` to
`ResponseHandler` refactoring.

lutovich committed:
Added rudimentary `Session#runAsync()`

It uses the existing simple async netty-based connection. Introduced a
simple async result cursor that only allows iteration over result
records. Also added a listenable future interface and implementation
which is an extension of default java future but allows listener
registration. Such future is needed because default java future only
gives ability to block for the result.

lutovich committed:
Initial work on netty channel pooling

Introduced an async connection pool that keeps pools of netty channels
by server address. It uses existing netty `FixedChannelPool`
internally. Also made `DirectConnectionProvider` use this pool and
acquire async connections from it.

New pool tries to expose API similar to exsiting `ConnectionPool`
interface except acquisition of connection returns an async connection
which is backed by a `Future<Channel>`.

Active connection tracking does not work because of
`ChannelInitializer` conflict in the channel pool.

lutovich committed:
Fixed tracking of active connection counts

Connection pool needs to track and expose active network connection
count for least-connected load balancing. Tracking and counts were
previously broken because of how `ChannelInitializer`s are handled
by the netty channel pool. Pool installs it's own channel initializer
which got overridden by our own initializer (which adds SSL and some
channel attributes). This prevented active channel tracking from
noticing a channel creation event and corresponding counter was
never decremented.

This commit makes our channel initializer aware of active connections
tracker which it notifies after new channel is initialized.

lutovich committed:
Small cleanup of channel future listeners

Extracted all listeners to separate top level classes and removed
unused methods from `AsyncConnection`.

lutovich committed:
Send RESET before releasing channel to the pool

To make sure database cleans up all resources and moves state machine
to the READY state.

lutovich committed:
Added async methods to StatementRunner

Fixed error handling and added couple more tests
for `Session#runAsync()`.

lutovich committed:
Fix synchronization when handling records

For each netty-based connection success, failure and record messages
are fetched from the network by an event loop thread. They are then
handed over to an application/user thread. This commit adds a separate
records handler and makes it synchronized to avoid any visibility
issues. Also added a simple TCP level back pressure with netty's
auto-read. When too many records are buffered event loop will not
read more from the network.

lutovich committed:
Respond to FAILURE with ACK_FAILURE

Bolt protocol requires clients to acknowledge failures by sending an
ACK_FAILURE message. Server will respond with IGNORED to every
subsequent request before failure is acknowledge. This commit makes
async connection eagerly send an ACK_FAILURE after receiving a FAILURE
message. It also improves fatal error handling. All queued response
handlers will be failed on fatal error. It's needed because fatal
errors are not recoverable and result in channel being closed.

lutovich committed:
Initial support of async transactions

This commit makes session able to start transactions asynchronously.
It is then possible to run async queries and async commit/rollback
those transactions. Async transactions currently do not track their
state and simply send issued commands to the server.

lutovich committed:
Further improve async transactions support

Fixed issues with connection release when async transactions are
committer/rolled back. Fused async transaction functionality into
`ExplicitTransaction`. Added `Session#closeAsync()`. Moved ACK_FAILURE
handling to the place where preceding FAILURE message is received,
ack was previously executed too late after user code.

lutovich committed:
Improved ListenableFuture

It is renamed to `Task` and has a dedicated listener class that
allows users to consume both result and failure.

lutovich committed:
Attempt to use futures bottom-up

This commit makes `Session#runAsync()` implementation use futures
all the way up. Previously it tried to use `NettyConnection` with
future inside. This was problematic because operations in the
corresponding event loop were not properly ordered.

It also simplifies connection release in result handler. Later now
directly releases connections without additional
abstraction `ResultResourcesHandler`.

lutovich committed:
Small cleanup of InternalStatementResultCursor

Made it take RUN and PULL_ALL handlers as constructor arguments.

lutovich committed:
Async transactions support based on futures

Reimplemented async transactions based on future chaining. This solves
issues with message queueing. All messages are queued and executed by
event loop which guarantees ordering.

lutovich committed:
Removed couple fixed TODOs

They were boltkit related and fixed by this commit
https://github.com/neo4j-contrib/boltkit/commit/730c6b2c7ec4bc183512111ea7b8ac7663aa363a.

lutovich committed:
Renamed internal package

From `netty` to `async` to better represent contained classes.

lutovich committed:
Clean up handling of outbound messages

Commit moves all classes related to netty outbound pipeline to a
separate package. It also make code properly handle messages that
do not fit in one chunk. Whole message is serialized to a netty
buffer primitive by primitive by `ChunkAwareByteBufOutput` which
also inserts chunk sizes where necessary.

lutovich committed:
Further simplify outbound message handling

Removed unnecessary abstraction OutboundMessageWriter.
OutboundMessageHandler now accepts message format which is selected
after handshake based on protocol version suggested by the server.

lutovich committed:
Clean up handling of inbound messages

Commit moves all classes related to netty inbound pipeline to a
separate package. It also adds unit tests for inbound handlers and
renames classes to better reflect their purpose.

lutovich committed:
Removed unused classes

lutovich committed:
Server version check for bytes, pool improvements

This commit includes:
 * switching off byte array support based on the server version after INIT
 * maxConnectionPoolSize & connectionAcquisitionTimeout are now propagated
   from Config to the pool
 * ActiveChannelTracker now keeps sets of all channels so that it can
   close them when needed; this functionality is currently unused but
   should be used for routing driver when updated routing table does
   not contain particular server
 * BootstrapFactory allows creation of single-threaded bootstraps for tests
 * server version is now attached to the channel attributes; it is currently
   unused but is valuable for routing driver to decide which routing
   procedure to invoke
 * more unit tests for various components

lutovich committed:
More unit tests and renaming

Renamed `EventLoopAwareFuture` to `InternalFuture` and
`EventLoopAwarePromise` to `InternalPromise`. Added unit tests for
connection initialization handlers/listeners.

lutovich committed:
Disallow commit a rolled back tx and vice versa

Such attempt will now return a failed future. Also fixed default
values for `maxConnectionPoolSize` and `connectionAcquisitionTimeout`.

lutovich committed:
Enforce connection timeout for netty channels

This commit makes channel bootstrap use and enforce connection timeout,
which is propagated from user through the config.

lutovich committed:
Implemented and removed couple minor TODOs

lutovich committed:
Cleanup around handling of futures

Named future combinators same as those in Java 8 `CompletableFuture`
and removed duplicated ones. Also these methods were added directly
to `InternalFuture` to improve code readability. Added convenience
method to convert `InternalFuture` to `Task` which will become part
of the public API.

lutovich committed:
Moved async pool related classes to separate package

lutovich committed:
More consistent method naming

`ActiveChannelTracker#prune()` renamed to `#purge()`.

lutovich committed:
Support long idle connection ping

This commit adds support for `idleTimeBeforeConnectionTest` with netty
channels. Ping/verification added to `NettyChannelHealthChecker` that
is used by the channel pool.

lutovich committed:
Expose statement keys in StatementResultCursor

Also this commit makes async query execution wait for RUN response
before returning cursor back to the user. This means all RUN failures
will surface earlier in a future returned by `#runAsync()` and not
during consumption of the cursor.

lutovich committed:
Expose result summary in StatementResultCursor

This commit makes result summary exposed as future.

lutovich committed:
Task renamed to Response and moved to public package

Also renamed related listener from `TaskListener` to `ResponseListener`.
Such renaming is needed because `Task` is more of a C# concept and
represents slightly different thing in Java.

Removed duplicated method from `InternalPromise`.

lutovich committed:
Fixed couple unit test failures

 * close async connection pool when driver creation fails
 * do not memorize transaction in session until it's successfully started
 * close async connection in `Session#close()`
 * relax assertion in connection timeout test because connection to
   non-routable IP addresses can result in immediate failure and not
   hanging in some environments

lutovich committed:
Renamed `InternalFuture#asTask()` to `#asResponse()`

zhenlineo commented:
LG, let's merge!

BTW, Should we merge this into a separate branch? or should we merge into main branch?

ali-ince commented on driver/src/main/java/org/neo4j/driver/internal/ConnectionSettings.java:
```diff
@@ -29,7 +29,7 @@
 public class ConnectionSettings
 {
-    private static final String DEFAULT_USER_AGENT = format( "neo4j-java/%s", driverVersion() );
+    public static final String DEFAULT_USER_AGENT = format( "neo4j-java/%s", driverVersion() );
```
Very small thing but is this being used elsewhere? Shall we set it as private again?

ali-ince commented on driver/src/main/java/org/neo4j/driver/internal/ExplicitTransaction.java:
```diff
@@ -57,28 +70,68 @@
+    {
+        Map<String,Value> parameters = initialBookmark.asBeginTransactionParameters();
+
+        connection.run( "BEGIN", parameters, NoOpResponseHandler.INSTANCE );
```
Do these "BEGIN", "COMMIT" and "ROLLBACK" strings  deserve their constants?

ali-ince reviewed (approved):
Looks really good! Only a couple of comments (even can be ignored :)).

lutovich committed:
Constants for BEGIN, COMMIT and ROLLBACK

lutovich committed:
Made DEFAULT_USER_AGENT a private constant

lutovich committed:
Simplify Response impl

By making `InternalFuture` implement this public interface. Previously
a special adapter object was needed. Now we can directly return
internal classes.

lutovich commented:
We decided to keep async API in a separate branch for now. It will get merged into 1.5 a bit later.
I'll close this PR now. Please track progress in https://github.com/neo4j/neo4j-java-driver/tree/1.5-async.
